design/one-pager-merge-jets-proposal.md

```diff
@@ -0,0 +1,402 @@
+  through Crossplane, i.e. coverage.
+* The providers should be mature enough for users to depend on them in
+  production, i.e. maturity.
+
+Today, we have two Crossplane providers for each of the big three clouds.
+* AWS
+  * Classic with [124 CRDs][provider-aws-crds] with 37 beta CRDs.
+  * Jet with [763 CRDs][jet-aws-crds], no beta CRDs.
+* Azure
+  * Classic with [21 CRDs][provider-azure-crds] with 7 beta CRDs.
+  * Jet with [647 CRDs][jet-azure-crds], no beta CRDs.
+* Google Cloud
+  * Classic with [27 CRDs][provider-gcp-crds] with 11 beta CRDs.
+  * Jet with [438 CRDs][jet-gcp-crds], no beta CRDs.
+
+Both variants of each reach those goals at different levels but none of them are
+maintained by their cloud vendor. The aim of this document is to capture a
+decision about the future of these provider implementations with the goals
+stated above in mind.
+
+## Proposal
+
+There were three options in the provider strategy doc:
+* Option A: Completely separate providers.
+* Option B: One provider but Jet-based resources in their own API group.
+* Option C: One provider and no Jet vs classic difference. Only API version
+  would be bumped if the implementation changes.
+
+We should go with the Option C, i.e. use Terrajet in
+[`provider-aws`][provider-aws], provider-azure and provider-gcp as another way
+of generating CRDs and controllers so that they can reach the goal of coverage
+really fast. Then the community can keep investing in that single provider and
+make each CRD mature enough to depend on in production. In this way, we can
+increase the usage by increasing both the coverage and the maturity of the CRDs
+in a consolidated way. This increase in usage can attract cloud vendors to come
+and join the maintainership of these providers. Once they do, they can either
+progressively convert the implementation of the controllers to their choice of
+technology or start with new providers which we can suggest people to use and we
+can provide migration tooling from the community-maintained one. In either case,
+the users would see a clear path forward and we'd not waste any efforts by the
+community.
+
+Note that we are making two decisions here:
+* There will be a single provider.
+* The single provider will be the classic one and we'll add Terrajet CRDs on
+  top.
+
+The order of execution will be roughly as following, taking
+[`provider-aws`][provider-aws] as an example:
+1. Add Terrajet pipeline to the provider repository.
+1. Copy the manual resource configurations from provider-jet-aws and introduce
+  the CRDs of those configurations in their new API group, without changing
+  their version, i.e. `v1alpha2`.
+1. Generate the full set of CRDs in `release-x.x-preview` branch similar to jet
+  repository and include those non-configured ones only in `-preview` suffixed
+  version.
+1. Write up a guide or tool to help the current users of non-preview version of
+  Jet providers to migrate to [`provider-aws`][provider-aws].
```

Comment:
The idea of migration is brilliant.
Besides, a smooth upgrade of the provider is essential for production users. Some of KubeVela uses classic Crossplane Alibaba Cloud CRDs in the business applications.

Reply:
The existing CRDs in classic providers will not be affected at all, we won't change their controller or schema with this plan today. The migration is mostly about users of Jet providers because once the merger is completed, those repositories will be archived so they will need to move, which will involve only API group change for most resources but also some schema changes if the CRD already existed in the classic provider (we don't replace the existing CRDs/controllers).